dali/base/dasds.cpp

```diff
@@ -2095,6 +2101,146 @@ ISDSManagerServer &querySDSServer()
+    mb.append(flags);
+    serializeVisibleAttributes(owner, mb);
+    StringBuffer extName;
+    getName(extName, name);
+    if (manager.extCache.lookup(extName, mb))
+        return;
+
+    if (withValue)
+    {
+        StringBuffer filename;
+        getFilename(filename, name);
+        Owned<IFile> iFile = createIFile(filename.str());
+        size32_t sz = (size32_t)iFile->size();
+        if ((unsigned)-1 == sz)
+        {
+            IptFlagClr(flags, ipt_binary);
+            mb.writeDirect(flagsPos, sizeof(flags), &flags);
+
+            StringBuffer s("Missing external file ");
+            if (*_name)
+                s.append("in property ").append(_name);
+            Owned<IException> e = MakeSDSException(SDSExcpt_MissingExternalFile, "%s", filename.str());
+            LOG(MCoperatorWarning, unknownJob, e, s.str());
+            StringBuffer str("EXTERNAL BINARY FILE: \"");
+            str.append(filename.str()).append("\" MISSING");
+            CPTValue v(str.length()+1, str.str(), false);
+            v.serialize(mb);
+        }
+        else
+        {
+            Owned<IFileIO> fileIO = iFile->open(IFOread);
+            verifyex(sz == ::read(fileIO, 0, sz, mb));
+            manager.extCache.add(extName, sz, mb.toByteArray());
+        }
+    }
+    else
+        mb.append((size32_t)0);
+}
+
+void CBinaryFileExternal::write(const char *name, IPropertyTree &tree)
+{
+    StringBuffer filename(name);
+    filename.append(ext);
+
+    MemoryBuffer out;
+    ((PTree &)tree).queryValue()->serialize(out);
+    size32_t len = out.length();
+    Owned<CTransactionItem> item = manager.deltaWriter.addExt(filename.detach(), len, out.detach());
+    manager.extCache.add(item);
+}
+
+/////////////////
+
+void CDeltaWriter::addToQueue(CTransactionItem *item)
+{
+    pending.push(item);
+    // add actual size for externals, and nonimal '100 byte' value for delta transactions
+    // it will act. as a rough guide to appoaching size threshold. It is not worth
+    // synchronously preparing and serializing here (which will be done asynchronously later)
+    pendingSz += (CTransactionItem::f_addext == item->type) ? item->dataLength : 100;
+    size_t items = pending.size();
+    if ((pendingSz < transactionMaxMem) && (items < transactionQueueLimit))
+    {
+        if (lastSaveTime && ((get_cycles_now() - lastSaveTime) < thresholdDuration))
+            return;
+
+        if (waiting)
+        {
+            waiting = false;
+            sem.signal();
+        }
+    }
+    else // here if exceeded transationQueueLimit, transactionMaxMem or exceeded time threshold (deltaSaveThresholdSecs)
+    {
+        ++totalQueueLimitHits;
+        // force a synchronous save
+        CCycleTimer timer;
+        CHECKEDCRITICALBLOCK(blockedSaveCrit, fakeCritTimeout); // because if Dali is saving state (::blockingSave), it will clear pending
+        PROGLOG("Forcing synchronous save of %u transactions", (unsigned)pending.size());
```

Comment:
Move before the critsec?  I suspect it is unlikely to be contended so not at all significant.

Reply:
but yes, logical to move before